File: programmers/lv3/섬_연결하기.py
parent: list


def find(x):
    if parent[x] < 0:
        return x

    parent[x] = find(parent[x])
    return parent[x]


def union(x, y):
    x = find(x)
    y = find(y)

    if x == y:
        return True

    if x > y:
        x, y = y, x

    parent[x] += parent[y]
    parent[y] = x


def solution(n, costs):
    global parent
    parent = [-1] * n
    costs.sort(key=lambda x: x[2])

    answer = 0
    cnt = 0
    for a, b, c in costs:
        if union(a, b):
            continue
        answer += c
        cnt += 1

        if cnt == n - 1:
            break

    return answer

File: programmers/lv3/섬_연결하기.java
import java.util.Arrays;
import java.util.Comparator;

class Solution {
    int[] parent;

    public int find(int x) {
        if (parent[x] < 0) return x;

        parent[x] = find(parent[x]);
        return parent[x];
    }

    public boolean union(int x, int y) {
        x = find(x);
        y = find(y);
        if (x == y) return false;

        if (parent[x] > parent[y]) {
            int temp = x;
            x = y;
            y = temp;
        }

        parent[x] += parent[y];
        parent[y] = x;

        return true;
    }

    public int solution(int n, int[][] costs) {

        parent = new int[n];
        Arrays.fill(parent, -1);

        Arrays.sort(costs, Comparator.comparing(cost -> cost[2]));

        int ans = 0;
        int cnt = 0;
        for (int[] cost : costs) {
            if (union(cost[0], cost[1])) {
                ans += cost[2];
                cnt++;

                if (cnt == n - 1) return ans;
            }
        }

        return ans;
    }
}